### PR TITLE
Use more hardware features in meson ethernet driver

### PR DESCRIPTION
**Pull request by Courtney3141**
This PR enables flow control and operate on second hardware features in the meson ethernet driver that improve UDP performance for the odroidc4. 

UDP throughput is improved for the odroidc4, TCP throughput is actually reduced, but with more consistency and lower CPU utilisation. It was decided that we would optimise for UDP performance for now.

### CODE REVIEW

File: drivers/network/meson/ethernet.c [hunk 1]
```diff
@@ -134,6 +137,7 @@ static void rx_return(void)
             rx.descr_mdata[rx.tail] = buffer;
             update_ring_slot(&rx, rx.tail, DESC_RXSTS_OWNBYDMA, cntl, buffer.io_or_offset, 0);
             eth_dma->rxpolldemand = POLL_DATA;
+            rx.tail = (rx.tail + 1) % RX_COUNT;
```
Comment (wom-bat) [inline drivers/network/meson/ethernet.c]: This change is unrelated.

File: drivers/network/meson/ethernet.c [hunk 2]
```diff
@@ -246,13 +250,23 @@ static void eth_setup(void)
...
 
     eth_dma->busmode = PRIORXTX_11 | ((DMA_PBL << TX_PBL_SHFT) & TX_PBL_MASK);
-    eth_dma->opmode = STOREFORWARD;
+    /*
+     * Operate in store-and-forward mode.
+     * Send pause frames when there's only 1k of fifo left,
+     * stop sending them when there is 2k of fifo left.
+     * Continue DMA on 2nd frame while updating status on first
+     */
+    eth_dma->opmode = STOREFORWARD | EN_FLOWCTL | (0 << FLOWCTL_SHFT) | (0 < DISFLOWCTL_SHFT) | TX_OPSCND;
```
Comment (wom-bat) [inline drivers/network/meson/ethernet.c:259]: Comment doesn't  match code.  Ideally it should be 1<< DISFLOWCTL_SHIFT to get the hysteresis (2k stop point)
From the Linux driver (RFA is the FLOWCTL_SHIFT; RFD DISFLOWCTL_SHIFT fields)

```
/* RFD and RFA fields are encoded as follows
 *
 *   Bit Field
 *   0,00 - Full minus 1KB (only valid when rxfifo >= 4KB and EFC enabled)
 *   0,01 - Full minus 2KB (only valid when rxfifo >= 4KB and EFC enabled)
 *   0,10 - Full minus 3KB (only valid when rxfifo >= 4KB and EFC enabled)
 *   0,11 - Full minus 4KB (only valid when rxfifo > 4KB and EFC enabled)
 *   1,00 - Full minus 5KB (only valid when rxfifo > 8KB and EFC enabled)
 *   1,01 - Full minus 6KB (only valid when rxfifo > 8KB and EFC enabled)
 *   1,10 - Full minus 7KB (only valid when rxfifo > 8KB and EFC enabled)
 *   1,11 - Reserved
 *
 * RFD should always be > RFA for a given FIFO size. RFD == RFA may work,
 * but packet throughput performance may not be as expected.
 *
 * Be sure that bit 3 in GMAC Register 6 is set for Unicast Pause frame
 * detection (IEEE Specification Requirement, Annex 31B, 31B.1, Pause
 * Description).
 *
 * Be sure that DZPA (bit 7 in Flow Control Register, GMAC Register 6),
 * is set to 0. This allows pause frames with a quanta of 0 to be sent
 * as an XOFF message to the link peer.
 */
 ```

File: drivers/network/meson/ethernet.h [hunk 3]
```diff
@@ -141,16 +141,32 @@
...
+#define DESC_TXCTRL_TXFIRST         (1 << 29)           /* Buffer contains the first segment of a frame. */
+#define DESC_TXCTRL_TXCRCDIS        (1 << 26)           /* GMAC does not append the Cyclic Redundancy Check (CRC) to the end of the transmitted frame.*/
+#define DESC_TXCTRL_TXRINGEND       (1 << 25)           /* Descriptor list reached its final descriptor. DMA must loop around. */
+#define DESC_TXCTRL_TXCHAIN         (1 << 24)           /* Second address in the descriptor is the Next Descriptor address rather than the second buffer address. */
+#define DESC_TXCTRL_SIZE2MASK       (0x3ff800)
+#define DESC_TXCTRL_SIZE2SHFT       (11)
+#define DESC_TXCTRL_SIZE1MASK       (0x7FF)
+#define DESC_TXCTRL_SIZE1SHFT       (0)
+
+/* GMAC FLOW CTRL defines */
```
Comment (wom-bat) [inline drivers/network/meson/ethernet.h:155]: These `GMAC_FLOW_CTRL` definitions repeat some earlier in the file.  We should decide which to keep, and adjust the places they're used accordingly.

**Review by wom-bat — CHANGES_REQUESTED** [review]

File: drivers/network/meson/ethernet.h (hunk 3 above)
Comment (Courtney3141) [reply drivers/network/meson/ethernet.h:155]: I think I'll move them up with the other definitions. Originally I designed the file so that the register definitions came first, followed by the descriptor definitions.

File: drivers/network/meson/ethernet.c (hunk 2 above)
Comment (Courtney3141) [reply drivers/network/meson/ethernet.c:259]: Yes, you're right, I missed that I fiddled around with different values when testing. I'll change the disable flow control value to Full minus 2K. Out of curiosity though I tested a few different configurations of these values and saw very little change in throughput from 756Mb/s and cpu utilisation ~ 80%. (Tried both (1K,1K), (1K,2K - what it is now), and (2K,3K).

**Review by wom-bat — APPROVED** [review]
LGTM now
